File: Software/Chapter 14/vb/Listing14-3/Listing14-3/Listing14-3.vb
Imports System
Imports System.Data
Imports System.Data.SqlClient
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server

Namespace APress.Samples
    Partial Public Class Sql
        <SqlProcedure()> _
        Public Shared Sub GetEnvironmentVars()
            Try
                Dim environmentList As New SortedList
                For Each de As DictionaryEntry In _
                    Environment.GetEnvironmentVariables()
                    environmentList(de.Key) = de.Value
                Next
                Dim record As New SqlDataRecord( _
                    New SqlMetaData("VarName", SqlDbType.NVarChar, 1024), _
                    New SqlMetaData("VarValue", SqlDbType.NVarChar, 4000))
                SqlContext.Pipe.SendResultsStart(record)
                For Each de As DictionaryEntry In environmentList
                    record.SetValue(0, de.Key)
                    record.SetValue(1, de.Value)
                    SqlContext.Pipe.SendResultsRow(record)
                Next
                SqlContext.Pipe.SendResultsEnd()
            Catch ex As Exception
                SqlContext.Pipe.Send(ex.Message)
            End Try
        End Sub
    End Class
End Namespace

File: Software/Chapter 14/cs/Listing14-3/Listing14-3/Listing14-3.cs
using System;
using System.Data;
using System.Data.SqlClient;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;
using System.Collections;

namespace APress.Samples
{
    public partial class Sql
    {
        [Microsoft.SqlServer.Server.SqlProcedure]
        public static void GetEnvironmentVars()
        {
            try
            {
                SortedList environmentList = new SortedList();
                foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
                {
                    environmentList[de.Key] = de.Value;
                }
                SqlDataRecord record = new SqlDataRecord(new SqlMetaData("VarName", SqlDbType.NVarChar, 1024),
                    new SqlMetaData("VarValue", SqlDbType.NVarChar, 4000));
                SqlContext.Pipe.SendResultsStart(record);
                foreach (DictionaryEntry de in environmentList)
                {
                    record.SetValue(0, de.Key);
                    record.SetValue(1, de.Value);
                    SqlContext.Pipe.SendResultsRow(record);
                }
                SqlContext.Pipe.SendResultsEnd();
            }
            catch (Exception ex)
            {
                SqlContext.Pipe.Send(ex.Message);
            }
        }
    };
}
